src/proto/network/pxe.rs

```diff
@@ -0,0 +1,1276 @@
+
+        let (buffer_ptr, mut buffer_size, dont_use_buffer) = if let Some(buffer) = buffer {
+            let buffer_ptr = NonNull::from(&mut buffer[0]).cast::<c_void>();
+            let buffer_size = u64::try_from(buffer.len()).unwrap();
+            (Some(buffer_ptr), buffer_size, false)
+        } else {
+            (None, 0, true)
+        };
+
+        let status = unsafe {
+            (self.mtftp)(
+                self,
+                TftpOpcode::TftpReadFile,
+                buffer_ptr,
+                false,
+                &mut buffer_size,
+                None,
+                server_ip,
+                Some(filename),
+                None,
+                dont_use_buffer,
+            )
+        };
+        Result::from(status)?;
+
+        Ok(buffer_size)
+    }
+
+    /// Writes to a file located on a TFTP server.
+    pub fn tftp_write_file(
+        &mut self,
+        server_ip: &IpAddress,
+        filename: &CStr8,
+        overwrite: bool,
+        buffer: &[u8],
+    ) -> Result {
+        let filename = NonNull::from(&filename.to_bytes_with_nul()[0]).cast();
+
+        let buffer_ptr = NonNull::from(&buffer[0]).cast();
+        let mut buffer_size = u64::try_from(buffer.len()).expect("buffer length should fit in u64");
+
+        unsafe {
+            (self.mtftp)(
+                self,
+                TftpOpcode::TftpWriteFile,
+                Some(buffer_ptr),
+                overwrite,
+                &mut buffer_size,
+                None,
+                server_ip,
+                Some(filename),
+                None,
+                false,
+            )
+        }
+        .into()
+    }
+
+    /// Reads a directory listing of a directory on a TFTP server.
+    pub fn tftp_read_dir<'a>(
+        &self,
+        server_ip: &IpAddress,
+        directory_name: &CStr8,
+        buffer: &'a mut [u8],
+    ) -> Result<impl Iterator<Item = TftpFileInfo<'a>> + 'a> {
+        let filename = NonNull::from(&directory_name.to_bytes_with_nul()[0]).cast();
+
+        let buffer_ptr = NonNull::from(&buffer[0]).cast();
+        let mut buffer_size = u64::try_from(buffer.len()).expect("buffer length should fit in u64");
+
+        let status = unsafe {
+            (self.mtftp)(
+                self,
+                TftpOpcode::TftpReadDirectory,
+                Some(buffer_ptr),
+                false,
+                &mut buffer_size,
+                None,
+                server_ip,
+                Some(filename),
+                None,
+                false,
+            )
+        };
+        Result::from(status)?;
+
+        let buffer_size = usize::try_from(buffer_size).expect("buffer length should fit in usize");
+        let buffer = &buffer[..buffer_size];
+
+        let mut iterator = buffer.split_inclusive(|b| *b == 0);
+
+        Ok(from_fn(move || {
+            let filename = iterator
+                .next()
+                .expect("the final entry should have an empty file name");
```

Comment:
Rather than calling `expect` here and below, could we change the type to `Result<TftpFileInfo, SomeNewErrorType>`? Akin to the [`ReadDir`](https://doc.rust-lang.org/std/fs/struct.ReadDir.html) iter. The error type doesn't necessarily need to be super fine-grained, mainly I just want to make sure that a bad ftp server doesn't cause this code to panic.